man/cqadm.8
.TH "cqadm" 8
.SH "NAME"
cqadm \- queue manager administration tool
.SH "SYNOPSIS"
.B cqadm [queue options] <queue> ... <queue>

.B cqadm [jobid options] <jobid> ... <jobid>

.B cqadm [no arg options]

.SH "DESCRIPTION"
Administrative interface for queue manager. Allows one to hold, run, kill, and set the queue for jobs in the queue manager. Can also add/del queues and set queue properties. 

.SH "OPTIONS WITH NO VALUES AND NO ARGUMENTS"

.TP
.B \-d \-\-debug
Turn on communication debugging
.TP
.B \-h \-\-help
Displays the usage and a brief descriptions of the options
.TP
.B \-\-version
Displays client revision and Cobalt version
.TP
.B \-f \-\-force
Force queue deletion.
.TP
.B \-\-getq
List queue(s) in queue manager

.SH "OPTIONS WITH VALUES BUT NO ARGUMENTS (Stand alone opions)"

.TP
.B \-j \-\-setjobid=<next jobid>
Set the jobID of the next job submitted to Cobalt
.TP
.B \-\-savestate=<filename>
Pickles all the components to the specified file

.SH "OPTIONS WITH NO VALUES BUT WITH QUEUE ARGUMENTS (Stand alone opions)"

.TP
.B \-\-addq queue(s)
Add specified queue(s) to queue manager
.TP
.B \-\-delq queue(s)
Delete specified queue(s) from queue manager
.TP
.B \-\-drainq queue(s)
Drain queue(s). No new jobs can be submitted, and jobs already in the queue will run normally. Queue state is set to 'draining'.
.TP
.B \-\-killq queue(s)
Kill the queue(s). No new jobs can be submitted, and no jobs in the queue will run. Queue state is set to 'dead'.
.TP
.B \-\-startq queue(s)
Start queue(s). Queue state is set to 'running'.
.TP
.B \-\-stopq queue(s)
Stop queue(s). Queue state is set to 'stopped'. New jobs can be submitted, but existing jobs will not be run.

.SH "OPTIONS WITH VALUES BUT WITH QUEUE ARGUMENTS (Stand alone options)"
.TP
.B \-\-policy <Queue Policy> queue(s)
Set queue policy for the queues specified.
.TP
.B \-\-setq 'prop=value prop=value' queue(s)
Set properties for one or more queues. The prop=value pairs are separated by whitespace.
.TP
.B \-\-unsetq 'property property' queue(s)
Move a property back to its unset state.  Used for example, to remove a queue restriction.

.SH "OPTIONS WITH NO VALUES BUT WITH JOBID ARGUMENTS (Stand alone options)"
.TP
.B \-\-kill jobid(s)
Kill job(s) using normal mechanism. Add -f or --force option for force kill. When forceable deletion used then this may leave processes dangling on computation resources.
.TP
.B \-\-preempt jobid(s)
Preempt jobid.
.TP
.B \-\-hold jobid(s) \- deprecated use \-\-admin\-hold instead
 Set job state to admin hold, preventing execution (only works on 'queued' jobs).
.TP
.B \-\-release jobid(s) \- deprecated use \-\-admin\-release instead
Set job that is in admin hold state to ready, allowing execution. 
.TP
.B \-\-admin\-hold jobid(s)
Set job state to admin hold, preventing execution (only works on 'queued' jobs)
.TP
.B \-\-admin\-release jobid(s)
Set job that is in admin hold state to ready, allowing execution.
.TP
.B \-\-user\-hold jobid(s)
Set job state to user hold, preventing execution (only works on 'queued' jobs)
.TP
.B \-\-user\-release jobid(s)
Set job that is in user hold state to ready, allowing execution

.SH "OPTIONS WITH VALUES and JOBID ARGUMENTS (The following options you can mix them together)"
.TP
.B \-\-queue=<queue> jobid(s)
Move job to specified queue (only works on jobs that are not 'running')
.TP
.B \-\-run=<location> jobid(s)
Run the job immediately on the specified location
.TP
.B \-\-time=<walltime> jobid(s)
Change job walltime to specified one
.TP
.SH "Properties"
.RS
.TP
.B maxtime=mm
The maximum walltime (in minutes) allowed for jobs submitted to the queue.
.TP
.B mintime=mm
The minimum walltime (in minutes) allowed for jobs submitted to the queue.
.TP
.B maxqueued=x
The maximum number of jobs a user is allowed to have in the queue, whether running or queued.
.TP
.B users=user1:user2:user2
List of users allowed to submit to the queue. An '*' means any user can submit to the queue.
.TP
.B state=running|stopped|draining|dead
Flag determining state of queue. If the queue is 'stopped', jobs can be submitted to the queue, but no jobs will be run by the scheduler. If the queue is 'draining', no new jobs will be allowed to be added to the queue, but jobs existing in the queue will be run. If the queue is 'dead', no new jobs will be added and no existing jobs will be run. (Also provided by 
.B "\-\-startq" "\-\-stopq" "\-\-drainq" "\-\-killq"
).
.TP
.B adminemail=user@example.com:nobody@example.com
Send an email at the start and stop of every job run through the specified queue.
.TP
.B maxrunning=x
The maximum number of jobs a user is allowed to have running in the queue.
.TP
.B maxtotaljobs=x
The maximum number of jobs a queue is allowed to run at once.
.TP
.B maxusernodes=x
The maximum number of nodes a user is allowed to have allocated with running jobs in the queue.
.TP
.B totalnodes=x
The maximum number of nodes any jobs in the queue are allowed to have allocated with running jobs.
.TP
.B priority=x
The priority that should be set for the queue.  The default priority for a queue is 0, and larger numbers indicate higher priority.

.SH "EXAMPLE"
\" .IP
Set no minimum walltime, a max walltime of 30 minutes, and only users 'mike, john, and george' can submit to someq:
.IP
.B cqadm --unsetq mintime someq
.IP
.B cqadm --setq 'maxtime=30 users=mike:john:george' someq


.RE
.SH "SEE ALSO"
.BR cqm(8)
.SH "BUGS"


CHANGES
= Changes from previous Cobalt Versions =

== Changes to 1.0.8 ==
 * Fixed a bug where a node going down during a running job could allow that
   node to be selected for draining.
 * Expanded a buffer used by the forkers that should significantly improve the
   speed of reading large ALPS responses and no longer interfere with node
   reacquisition
 * Cleaned up stray print statements in node information
 * Fixed an error where score could be set to a string rather than a float via
   the filter scripts
 * Fixed an error where the queue-manager statefile could be lost if Cobalt was
   brought down with a job in the "Terminal" state.

== Changes to 1.0.6 ==
 * Backfill now supported on Cray systems.
 * Major performance improvement to nodelist and nodeadm -l.
 * Fix for an error that can cause loss of job tracking during job startup
   resulting in a reported exit status of "1234567".
 * Fix for an error in system_script_forker that can cause an error loop
   if using stdout to string functionality.
 * Fix for an error that caused jobs with a "terminating" status to cause
   qstat to fail.

== Changes to 1.0.5 ==
 * Fix for an issue that would cause jobs with their location set to not
   run if they were submitted to a reservation.
 * Performance of setres improved significantly.
 * Fixed an issue where reservations and soft-down status on nodes may not
   be respected by job placement.
 * An enhanced_apkill script is now available to improve job cleanup and
   ensure that jobs that ignore signals are otherwise properly terminated.

== Changes to 1.0.4 ==
 * Fix for apid not being properly checked when apkill was being invoked.

== Changes to 1.0.3 ==
 * Fix for an issue for nodes coming out of disabled state and suddenly
   appearing in ALPS.
 * Interactive job support
 * Support for CAPMC memory management scripts added.  Cobalt now gracefully
   handles long startup times if these are being used.
 * Improved job cleanup using apkill.

== Changes to 1.0.2 ==
 * Singleton job filter (maxtotaljobs) now supported
 * Multiple alps_script_forkers running simultaneously is now supported. Forkers
   are chosen for execution based on which forker has the lowest number of jobs
   currently assigned to it.

== Changes to 1.0.0 ==
 * Initial support of Cray systems

== Changes to 0.99.42 ==
 * Update to database writer to handle attrs fields that are
   improperly formatted.

== Changes to 0.99.41 ==
 * Minor patch to clean up an issue where the user was not being properly
   cleared from BlueGene blocks at the end of a run.

== Changes to 0.99.40 ==
 * A nokillpg attribute may be added to a job to cause kill instead of killpg
   to be used for job termination so that a signal handler in a script may
   handle signal propigation itself.
 * Fixed a bug where a job requesting a location could cause Cobalt to drain
   for a job that would require the use of offline hardware
 * Added a check to the job validator where a job is rejected if the job size
   and block geometry have a mismatch.

== Changes to 0.99.39 ==
 * Session information added to job submission information.  This also includes
   submitting host information (Cobalt Ticket #4315)
 * A fix for a bug that can cause "phantom drains" to occur on lightly loaded
   cluster systems. (Cobalt Ticket #4319)

== Changes to 0.99.38 ==
 * Added numerous dependency changes.  Notably, jobs in dep_hold states will no
   longer accrue eligible wait time or score.  Previously, dep_hold jobs would
   accrue eligible wait times unlike admin_hold or user_hold states.
   (Cobalt Ticket #4308)
 * Two display fixes for cluster systems that allow multiple node-name prefixes
   to be handled correctly by the nodelist merging code  and to correctly
   compact names if leading 0's are used in location numbers like
   "foo001.machine". (Cobalt Ticket #4313 and #4277)
 * Corrected two race conditions in the BlueGene/Q system components where a
   location with offline hardware could be selected for draining on a
   scheduling pass.  Though Cobalt would never try to actually run on the
   offline block, It could end up draining resources needlessly, and impact
   machine usage.  The first could happen at any time, the second only occurred
   at the end of booting a block. (Cobalt Ticket #4311)

== Changes to 0.99.37 ==
 * Fixed issue where a reservation could hit it's activation check while it is being
   deactivated, causing those events should be reversed.  Released reservations can't
   activate. (Cobalt Ticket #4289)
 * COBALT_JOBSIZE and COBALT_PARTSIZE have been added to interactive jobs.
   (Cobalt Ticket #4294)
 * Job dependencies require the same project to gain score from eachother.
   (Cobalt Ticket #4293)
 * prologue_helper.py now handles "=" in arguments correctly. (Cobalt Ticket #4292)

== Changes to 0.99.36 ==
 * Fix for issues in get-bootable-blocks and with block booting.  Block names are now
   properly sorted.  A race condition between boot-block and update_block_state has
   been fixed.  The return statuses for boot-block have also been corrected for failed
   boots.  (Cobalt Ticket #4282 and #4280)

== Changes to 0.99.35 ==
 * Fixed an issue where a block being removed in the BGQ control system between block
   detection and the block being added for tracking could result in blocks being
   erroniously marked as idle. (Cobalt Ticket #4268)

== Changes to 0.99.34 ==
 * Fixed an issue where Cobalt could have a memory leak if boot-block.py was
   terminated early the boot tracking object may never be cleaned up.
 * Numerous enhancements and fixes to handle multiple overlaping queues in the
   system with respect to draining and backfilling. (Cobalt Ticket #663)
 * Fixed some typos in the documentation for qsub.
 * COBALT_RESID is now appropriately set in interactive mode jobs (Cobalt Ticket #703)

== Changes to 0.99.33 ==
 * Fixed issue where adding blocks while the system component is running can

src/clients/cqadm.py
#!/usr/bin/env python
"""
Administrative interface for queue manager. Allowsone to hold, run, kill, and set 
the queue for jobs in the queue manager. Can also add/del queues and set queue properties.

Usage: %prog --help
Usage: %prog [options] <jobid> <jobid> OR <queue> <queue>
version: "%prog " + __revision__ + , Cobalt  + __version__

OPTIONS DEFINITIONS:

Options with no values and no arguments:

'-d','--debug',dest='debug',help='turn on communication debugging',callback=cb_debug
'-f','--force',dest='force',default=False,help='force flag used by some options',action='store_true'
'--getq',dest='getq',help='get information for all queues',action='store_true'

Options with values but no arguments:

'-j','--setjobid',dest='setjobid',type='int',help='set next jobid',callback=cb_gtzero
'--savestate',dest='savestate',type='string',help='save state to specified filename',callback=cb_path

Options with no values but with queue arguments:

'--addq',dest='addq',help='add queues in args',action='store_true'
'--delq',dest='delq',help='delete queues in args',action='store_true'
'--drainq',help='drain queues in args',callback=cb_setqueues
'--killq',help='kill queues in args',callback=cb_setqueues
'--startq',help='start queues in args',callback=cb_setqueues
'--stopq',help='stop queues in args',callback=cb_setqueues

Option with values and queue arguments:

'--policy',dest='qdata',type='string',help='set policy for queues in args',callback=cb_setqueues

'--setq',dest='qdata',type='string',help='props for queues in args ("prop1=val1 prop2=val2 ... propN=valN")', /
  callback=cb_setqueues

'--unsetq',dest='qdata',type='string',help='unset props for queues in args ("prop1 prop2 ... prop3")', /
  callback=cb_setqueues

Options with no values but with jobid arguments:

'--kill',dest='kill',help='kill jobids in args',action='store_true'
'--preempt',dest='preempt',help='preempt jobids in args',action='store_true'
'--hold', help='Deprecated use --admin-hold instead', callback=cb_hold
'--release', help='Deprecated use --admin-release instead', callback=cb_hold
'--admin-hold', help='apply admin hold to jobids in args',callback=cb_hold
'--admin-release', help='release admin hold for jobids in args', callback=cb_hold
'--user-hold', help='apply user hold to jobids in args', callback=cb_hold
'--user-release', help='release user hold for jobids in args', callback=cb_hold

Option with values and jobid arguments:

'--queue',dest='queue',type='string',help='modify queue name for jobids in args'
'--run',dest='run',type='string',help='modify run location for jobids in args'

'--time',dest='walltime',type='string',help='modify walltime for jobids in args (minutes or HH:MM:SS)', /
  callback=cb_time

"""
import logging
import sys
from Cobalt import client_utils
from Cobalt.client_utils import \
    cb_debug, cb_time, cb_path, cb_gtzero, cb_setqueues, cb_hold
from Cobalt.arg_parser import ArgParse

    
__revision__ = '$Revision: 559 $'
__version__  = '$Version$'

SYSMGR = client_utils.SYSMGR
QUEMGR = client_utils.QUEMGR

def check_option_conflicts(opt_count, parser):
    """
    check mutually exclusive options
    """
    errmsg = '' # init error msessage to empty string
    # Check mutually exclusive options
    if opt_count > 1:
        if parser.options.setjobid  != None: 
            errmsg += ' setjobid'
        if parser.options.savestate != None: 
            errmsg += ' savestate'
        if parser.options.run       != None: 
            errmsg += ' run'
        if parser.options.addq      != None: 
            errmsg += ' addq'
        if parser.options.getq      != None: 
            errmsg += ' getq'
        if parser.options.delq      != None: 
            errmsg += ' delq'
        if parser.options.qdata     != None: 
            errmsg += ' ' + parser.options.setq_opt # set queue options
        if parser.options.preempt   != None: 
            errmsg += ' preempt'
        if parser.options.kill      != None: 
            errmsg += ' kill'

    if errmsg != '':
        errmsg = 'Option combinations not allowed with: %s option(s)' % errmsg[1:].replace(' ', ', ')
        client_utils.logger.error(errmsg)
        sys.exit(1)


def validate_args(parser, spec, opt_count):
    """
    Validate qalter arguments
    """
    opts_wo_args = ['debug', 'getq', 'savestate', 'setjobid'] # no argument options

    # handle release or hold options
    if hasattr(parser.options, 'admin_hold'):
        opt_count += 1
        spec['admin_hold'] = parser.options.admin_hold

    if hasattr(parser.options, 'user_hold'):
        opt_count += 1
        spec['user_hold'] = parser.options.user_hold
    
    # Make sure jobid or queue is supplied for the appropriate commands
    if parser.no_args() and not [opt for opt in spec if opt in opts_wo_args]:
        client_utils.print_usage(parser)
        sys.exit(1)

    # Check required options
    if opt_count == 0:
        client_utils.logger.error("At least one option must be specified")
        sys.exit(1)

    check_option_conflicts(opt_count, parser)

    if parser.options.addq   != None or \
       parser.options.delq   != None or \
       parser.options.getq   != None or \
       parser.options.qdata  != None: # set queue options
        # queue job change request
        jobs = [{'tag':'queue', 'name':qname} for qname in parser.args]
    else:
        # get jobids from the argument list
        jobids = client_utils.get_jobids(parser.args)
        jobs = [{'tag':'job', 'jobid':int(jobid), 'location':'*', 'walltime':'*'} for jobid in jobids]

    return jobs

def getq(info):
    """
    get queue
    """
    response = client_utils.component_call(QUEMGR, True, 'get_queues', (info,))
    for que in response:
        if que['maxtime'] is not None:
            que['maxtime'] = "%02d:%02d:00" % (divmod(int(que.get('maxtime')), 60))
        if que['mintime'] is not None:
            que['mintime'] = "%02d:%02d:00" % (divmod(int(que.get('mintime')), 60))
    header = [('Queue', 'Users', 'Groups', 'MinTime', 'MaxTime', 'MaxRunning',
                'MaxTotalJobs', 'MaxQueued', 'MaxUserNodes', 'MaxNodeHours',
                'TotalNodes', 'AdminEmail', 'State', 'Cron', 'Policy', 'Priority')]
    datatoprint = [(que['name'], que['users'], que['groups'],
                    que['mintime'], que['maxtime'],
                    que['maxrunning'],  que['maxtotaljobs'], que['maxqueued'],
                    que['maxusernodes'], que['maxnodehours'],
                    que['totalnodes'],
                    que['adminemail'], que['state'],
                    que['cron'], que['policy'], que['priority'])
                   for que in response]
    datatoprint.sort()
    client_utils.print_tabular(header + datatoprint)
    return response

def setjobs(jobs, parser, spec, user):
    """
    set jobs 
    """
    if hasattr(parser.options,'admin_hold'):
        for job in jobs:
            job.update({'admin_hold': not parser.options.admin_hold})

    if hasattr(parser.options,'user_hold'):
        for job in jobs:
            job.update({'user_hold': not parser.options.user_hold})

    return client_utils.component_call(QUEMGR, False, 'set_jobs', (jobs, spec, user))

def process_cqadm_options(jobs, parser, spec, user):
    """
    This function will process any command argument and options passed to cqadm
    """

    force = parser.options.force # force flag. 

    info = [{'tag':'queue', 'name':'*', 'state':'*', 'users':'*', 'groups':'*', 'maxtime':'*', 'mintime':'*', 'maxuserjobs':'*',
             'maxusernodes':'*', 'maxqueued':'*', 'maxrunning':'*','maxtotaljobs':'*', 'maxnodehours':'*', 'adminemail':'*', 
             'totalnodes':'*', 'cron':'*', 'policy':'*', 'priority':'*'}]

    response = []
    if parser.options.setjobid != None:
        response = client_utils.component_call(QUEMGR, True, 'set_jobid', (parser.options.setjobid, user))

    elif parser.options.savestate != None:
        response = client_utils.component_call(QUEMGR, True, 'save', (parser.options.savestate,))

    elif parser.options.kill != None:
        response = client_utils.component_call(QUEMGR, False, 'del_jobs', (jobs, force, user))

    elif parser.options.run != None:
        response = client_utils.run_jobs(jobs, parser.options.run, user)

    elif parser.options.addq != None:
        response = client_utils.add_queues(jobs, parser, user, info)

    elif parser.options.getq != None:
        response = getq(info)

    elif parser.options.delq != None:
        response = client_utils.del_queues(jobs, force, user)

    elif parser.options.qdata != None:
        response = client_utils.component_call(QUEMGR, True, 'set_queues', (jobs, parser.options.qdata, user))

    elif parser.options.preempt != None:
        response = client_utils.component_call(QUEMGR, True, 'preempt_jobs', (jobs, user, force))

    else:
        response = setjobs(jobs, parser, spec, user)

    if not response:
        client_utils.logger.error("Failed to match any jobs or queues")
    else:
        client_utils.logger.debug(response)

def main():
    """
    cqadm main function.
    """
    # setup logging for client. The clients should call this before doing anything else.
    client_utils.setup_logging(logging.INFO)

    spec     = {} # map of destination option strings and parsed values
    opts     = {} # old map
    opt2spec = {}

    # list of callback with its arguments
    callbacks = [
        # <cb function>     <cb args (tuple) >
        ( cb_debug        , () ),
        ( cb_gtzero       , (True,) ), # return int
        ( cb_time         , (False, False, False) ), # no delta time, minutes, return string
        ( cb_path         , (opts, False) ), # do not use CWD
        ( cb_setqueues    , () ),
        ( cb_hold         , () ) ]

    # Get the version information
    opt_def =  __doc__.replace('__revision__', __revision__)
    opt_def =  opt_def.replace('__version__', __version__)

    parser = ArgParse(opt_def, callbacks)

    user = client_utils.getuid()

    # Set required default values: None

    parser.parse_it() # parse the command line
    opt_count = client_utils.get_options(spec, opts, opt2spec, parser)

    jobs = validate_args(parser, spec, opt_count)

    process_cqadm_options(jobs, parser, spec, user)

if __name__ == '__main__':
    try:
        main()
    except SystemExit:
        raise
    except Exception, e:
        client_utils.logger.fatal("*** FATAL EXCEPTION: %s ***", e)
        sys.exit(1)
